src/data/string/ustring.h

```diff
@@ -75,72 +76,121 @@ uchar string_to_uchar(const char* p);
 uchar string_to_uchar(const std::string& s);
 std::string uchar_to_string(uchar uc);
 
+class fallback_base {
+public:
+  virtual ~fallback_base() {
+  }
+
+  virtual uchar fallback(const std::string& bytes,
+			 const char* hint) = 0;
+};
+
+class exception_fallback : public fallback_base {
+public:
+  virtual ~exception_fallback() {
+  }
+
+  virtual uchar fallback(const std::string& bytes,
+			 const char* hint) {
+    throw std::invalid_argument(hint);
+    return 0x0000;
+  }
+};
+
+class replacement_fallback : public fallback_base {
+public:
+  replacement_fallback(uchar replace) : replace(replace) {
+  }
+
+  virtual ~replacement_fallback() {
+  }
+
+  virtual uchar fallback(const std::string& bytes,
+			 const char* hint) {
+    if (bytes.empty()) {
+      return 0x0000;
+    }
+    return replace;
+  }
+
+private:
+  uchar replace;
+};
+
 namespace detail {
 template <class InputIterator1, class InputIterator2>
-uchar chars_to_uchar_impl(InputIterator1& in, InputIterator2 end)
+uchar chars_to_uchar_impl(InputIterator1& in, InputIterator2 end,
+			  fallback_base& fb)
 {
+  InputIterator1 begin = in;
```

Comment:
This `begin` is always point to the original head of argument, and never change in this scope.
`const` might make it more clear, i think.